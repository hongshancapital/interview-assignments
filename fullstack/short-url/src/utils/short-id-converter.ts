import Hashids from 'hashids'

export type Id = number
export type HashId = string

const base58alphabet = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'

const hashids = new Hashids("short-url", 8, base58alphabet);

export const encode = (id: Id) : HashId => hashids.encode(id)
export const decode = (hashId: HashId) => hashids.decode(hashId)[0]