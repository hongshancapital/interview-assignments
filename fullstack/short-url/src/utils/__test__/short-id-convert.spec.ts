import * as conv from '../short-id-converter'

describe('short id convert', () => {
  it('should return fixed uniq 8 length short id when encode give any int number <= 194754273880', async () => {
    const hashIdMin = conv.encode(0)
    const hashIdMax = conv.encode(194754273880)
    const hashIdOver = conv.encode(194754273881)

    expect(hashIdMin).toHaveLength(8)
    expect(hashIdMax).toHaveLength(8)

    expect(hashIdMin).toMatchSnapshot()
    expect(hashIdMax).toMatchSnapshot()

    expect(hashIdMin).not.toEqual(hashIdMax)
    expect(hashIdOver).toHaveLength(9)
  });

  it('should return original number id when decode give short id', async () => {
    const hashIdOne = conv.encode(3)
    const hashIdTwo = conv.encode(4)

    expect(conv.decode(hashIdOne)).toEqual(3)
    expect(conv.decode(hashIdTwo)).toEqual(4)
  });
});
