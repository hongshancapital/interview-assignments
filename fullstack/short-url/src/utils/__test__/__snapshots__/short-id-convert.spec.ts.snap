// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`short id convert should return fixed uniq 8 length short id when encode give any int number <= 194754273880 1`] = `"XJA9v3m8"`;

exports[`short id convert should return fixed uniq 8 length short id when encode give any int number <= 194754273880 2`] = `"Wnnnnnnn"`;
